### ISSUE TITLE
Metrics not found when deleted and recreated with experiment

### ISSUE THREAD

**Issue by ukclivecox** [iter8-tools/iter8#782]
**Describe the bug**
Run an experiment with metrics once. Delete and rerun immediately. Sometimes it will fail with metric not found.

**To Reproduce**

Create Kind cluster
```
kind create cluster
```

Setup platform

```
cd ${ITER8}/samples/seldon/quickstart
./platformsetup.sh
```

Run code below which start experiment/metrics 3 times in a row.

```
#!/bin/bash

set -e

export EXPERIMENT=quickstart-exp

echo "Create app/ML model versions"
kubectl apply -f $ITER8/samples/seldon/quickstart/baseline.yaml
kubectl apply -f $ITER8/samples/seldon/quickstart/candidate.yaml
kubectl apply -f $ITER8/samples/seldon/quickstart/routing-rule.yaml
kubectl wait --for condition=ready --timeout=600s pods --all -n ns-baseline
kubectl wait --for condition=ready --timeout=600s pods --all -n ns-candidate

echo "Generate requests"
URL_VALUE="http://$(kubectl -n istio-system get svc istio-ingressgateway -o jsonpath='{.spec.clusterIP}'):80"
sed "s+URL_VALUE+${URL_VALUE}+g" $ITER8/samples/seldon/quickstart/fortio.yaml | sed "s/6000s/600s/g" | kubectl apply -f -
pod_name=$(kubectl get pods --selector=job-name=fortio-requests -o jsonpath='{.items[*].metadata.name}')
kubectl wait --for=condition=Ready pods/"$pod_name" --timeout=240s
pod_name=$(kubectl get pods --selector=job-name=fortio-irisv1-rewards -o jsonpath='{.items[*].metadata.name}')
kubectl wait --for=condition=Ready pods/"$pod_name" --timeout=240s
pod_name=$(kubectl get pods --selector=job-name=fortio-irisv2-rewards -o jsonpath='{.items[*].metadata.name}')
kubectl wait --for=condition=Ready pods/"$pod_name" --timeout=240s

echo "Define metrics"
kubectl apply -f $ITER8/samples/seldon/quickstart/metrics.yaml

echo "Launch experiment"
kubectl apply -f $ITER8/samples/seldon/quickstart/experiment.yaml


# Wait a few seconds
sleep 10

kubectl delete -f $ITER8/samples/seldon/quickstart/experiment.yaml
kubectl delete -f $ITER8/samples/seldon/quickstart/metrics.yaml

echo "Define metrics"
kubectl apply -f $ITER8/samples/seldon/quickstart/metrics.yaml

echo "Launch experiment"
kubectl apply -f $ITER8/samples/seldon/quickstart/experiment.yaml

sleep 5

kubectl get experiment $EXPERIMENT -o yaml

# Wait a few seconds
sleep 10

kubectl delete -f $ITER8/samples/seldon/quickstart/experiment.yaml
kubectl delete -f $ITER8/samples/seldon/quickstart/metrics.yaml

echo "Define metrics"
kubectl apply -f $ITER8/samples/seldon/quickstart/metrics.yaml

echo "Launch experiment"
kubectl apply -f $ITER8/samples/seldon/quickstart/experiment.yaml

sleep 5

kubectl get experiment $EXPERIMENT -o yaml
```

**Expected behavior**
Metrics are found.

**Kubernetes stack (please provide the following information):**
 - K8s environment: Kind
 - K8s version 1.19
 - K8s stack:  istio/seldon

I imagine the problem could be due to caching / stale objects in watchers in operator for metrics.

**Comment by sriumcp** [comment]
`kubectl apply` returns after successfully submitting objects to the K8s API server (not after the server has successfully created/changed the resources). Couldn't find a proper reference for this except: https://github.com/kubernetes/kubernetes/issues/72817#issuecomment-453709264
***
So, during platform setup, `kubectl apply metrics` is successfully submitted, and then `kubectl apply experiment` is successfully submitted, and then the experiment begins before the API server has finished creating metrics.

This race condition might explain the failures above, and also why the failures occur sometimes but not always.

**Comment by sriumcp** [comment]
Given that the declarative `kubectl apply` is the [recommended approach](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#kubectl-apply)  to manage K8s production clusters rather than the imperative `kubectl create` (which I am guessing blocks until creation is complete), there are a few things that come to mind.

1. **[Quick hack]** Introduce a sleep (example 5 seconds) between `kubectl apply metrics` and `kubectl apply experiment` in the script. Obviously not clean, but would make the problem go away most of the time in this script (it really shouldn't take the K8s API server 5 seconds to create metrics, though it could -- when it is heavily loaded or resource constrained or stuck in a bad state).
***
2. **[A better approach]** Write an Iter8 task which will wait until all resources referenced in the experiment are available.
```yaml
# example
- start: 
  - task: common/readiness
    with: #optional
      timeout: 60s # any go duration string; optional; default = 60s
```
> if objects referenced in the experiment are unavailable within 60 seconds, the above task, and therefore the experiment will fail with an appropriate error message/failure condition.
***
3. **[An even better approach; I will create this feature request shortly]** Extend the above task so that it actually waits until resources are ready, not just available. For example, allow specification of any readiness condition for any resource (not just those referenced in the experiment object) as part of its input.
```yaml
# example
- start: 
  - task: common/readiness
    with: # optional
      timeout: 60s # any go duration string; optional; default = 60s
      resources: # optional
      - objRef: pod/busybox1 # required
        condition: Ready # required
        namespace: default #optional
```

**Note:** Any condition that you may specify in a `kubectl wait` command is allowed in the condition field above. `kubectl wait` doesn't check for the existence of a resource; only the given condition in the object's status. The Iter8 `metrics` object does not have a status field in the first place... so `kubectl wait` is useless as far as metrics are concerned.

**Note:** Waiting for readiness of resources involves RBACs. So, the required RBAC permissions are either available as part of Iter8 installation (metrics example), or supplied by the user (pod example). Without appropriate RBACs, the task will fail and so will the experiment.

**Comment by sriumcp** [comment]
Incidentally, currently available Iter8 tasks are documented here. https://iter8.tools/0.6/reference/tasks/

Some tasks are involved, but many are easy to write and can leverage existing tools like `kubectl` under the covers as part of their implementation. Perhaps we also put together a document on how to author a task using a simple example (the `common/readiness` task is in fact a really simple, short, and sweet task and could serve as a good candidate for this documentation).

Will create an issue for this separately.

**Comment by sriumcp** [comment]
Before implementing `common/readiness`, I will use the **quick hack approach** to verify that the problem is indeed the race condition described above.
***
Regardless, the above mentioned race condition is certainly an issue. It can become pronounced in GitOps contexts (the GitOps operator can take its own sweet time to apply resources). 

For these reasons, it may be useful to include the `common/readiness` task (without `resources` stanza) in every Iter8 experiment, and optionally include the `resources` stanza, depending on the specific experiment.

**Comment by ukclivecox** [comment]
> `kubectl apply` returns after successfully submitting objects to the K8s API server (not after the server has successfully created/changed the resources). Couldn't find a proper reference for this except: [kubernetes/kubernetes#72817 (comment)](https://github.com/kubernetes/kubernetes/issues/72817#issuecomment-453709264)
> 
> So, during platform setup, `kubectl apply metrics` is successfully submitted, and then `kubectl apply experiment` is successfully submitted, and then the experiment begins before the API server has finished creating metrics.
> 
> This race condition might explain the failures above, and also why the failures occur sometimes but not always.

Yes would be good to confirm its due to `apply` rather than `create` as opposed to something more deep. I've not yet seen it on initial Metrics creations rather than recreation which would imply something deeper.

**Comment by sriumcp** [comment]
Here is a run that replicates the issue... https://github.com/iter8-tools/iter8/runs/2822565309?check_suite_focus=true

**Comment by sriumcp** [comment]
Here is a run that verifies that the quick hack works... and settles the race condition hypothesis! : https://github.com/iter8-tools/iter8/runs/2822616555?check_suite_focus=true

**Comment by sriumcp** [comment]
@cliveseldon @kalantar FYI.

**Comment by ukclivecox** [comment]
Great

**Comment by ukclivecox** [comment]
@sriumcp Maybe I missed it but what is the quick hack to fix this?

**Comment by sriumcp** [comment]
@cliveseldon see [here](https://github.com/iter8-tools/iter8/issues/782#issuecomment-860042734)

**TL;DR** 

**[Quick hack]** Introduce a sleep (example 5 seconds) between `kubectl apply metrics` and `kubectl apply experiment` in the script. Obviously not clean, but would make the problem go away most of the time in this script.

**Comment by ukclivecox** [comment]
> @cliveseldon see [here](https://github.com/iter8-tools/iter8/issues/782#issuecomment-860042734)
> 
> **TL;DR**
> 
> **[Quick hack]** Introduce a sleep (example 5 seconds) between `kubectl apply metrics` and `kubectl apply experiment` in the script. Obviously not clean, but would make the problem go away most of the time in this script.

OK. I've certainly seen more than 5 secs. Did you check its not stale information in the operator?

**Comment by sriumcp** [comment]
There is no stale information in the Iter8 controller as far as metrics are concerned... this is because the controller does not watch for metrics at all. It actively fetches metrics (equivalent to `kubectl get` but using the K8s go client) each time an experiment is created and populates the metrics field in the experiment status with metric definitions.

https://github.com/iter8-tools/etc3/blob/f5a34464b7cedfabdd023868bea153e9c57a0217/controllers/experiment_controller.go#L170

So, either metrics are present when experiment is created (experiment moves ahead), or not present (experiment fails; the case you are describing).

I agree `5s` is unreliable (which is precisely  why this is a quick hack; we can increase `5s` to `20s`, perhaps less glitchy but definitely just as hacky!)

Also,  note that the only difference between the [failed runs](https://github.com/iter8-tools/iter8/runs/2822565309?check_suite_focus=true) and the [successful runs](https://github.com/iter8-tools/iter8/runs/2822616555?check_suite_focus=true) above is  the `5s` sleep.
